File: 3630.partition-array-for-maximum-xor-and-and.py
#
# @lc app=leetcode id=3630 lang=python3
#
# [3630] Partition Array for Maximum XOR and AND
#

# @lc code=start
class Solution:
    def maximizeXorAndXor(self, nums: List[int]) -> int:
        n = len(nums)
        max_result = 0
        for i in range(1 << n):
            A, B, C = [], [], []
            for j in range(n):
                if i & (1 << j):
                    A.append(nums[j])
                elif (i >> j) & 1 == 0:
                    B.append(nums[j])
                else:
                    C.append(nums[j])
            xor_A = 0 if not A else reduce(lambda x, y: x ^ y, A)
            and_B = (1 << 32) - 1 if not B else reduce(lambda x, y: x & y, B)
            xor_C = 0 if not C else reduce(lambda x, y: x ^ y, C)
            max_result = max(max_result, xor_A + and_B + xor_C)
        return max_result # @lc code=end

File: 3630.partition-array-for-maximum-xor-and-and.go
#
# @lc app=leetcode id=3630 lang=golang
#
# [3630] Partition Array for Maximum XOR and AND
#
# @lc code=start
func maximizeXorAndXor(nums []int) int64 {
    n := len(nums)
    type key struct {
        i, xorA, andB, xorC int
    }
    memo := make(map[key]int64)
    var dfs func(i int, xorA int, andB int, xorC int) int64
    dfs = func(i int, xorA int, andB int, xorC int) int64 {
        k := key{i, xorA, andB, xorC}
        if val, exists := memo[k]; exists {
            return val
        }
        if i == n {
            ab := 0
            if andB != -1 {
                ab = andB
            }
            return int64(xorA + ab + xorC)
        }
        // Assign nums[i] to A
        resA := dfs(i+1, xorA^nums[i], andB, xorC)
        // Assign nums[i] to B
        newAndB := nums[i]
        if andB != -1 {
            newAndB = andB & nums[i]
        }
        resB := dfs(i+1, xorA, newAndB, xorC)
        // Assign nums[i] to C
        resC := dfs(i+1, xorA, andB, xorC^nums[i])
        maxRes := resA
        if resB > maxRes {
            maxRes = resB
        }
        if resC > maxRes {
            maxRes = resC
        }
        memo[k] = maxRes
        return maxRes
    }
    return dfs(0, 0, -1, 0)
}
# @lc code=end